pkg/apis/certificates/v1beta1/defaults.go

```diff
@@ -17,15 +17,73 @@ limitations under the License.
 package v1beta1
 
 import (
+	"crypto/x509"
+	"reflect"
+	"strings"
+
 	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
 )
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
+
 func SetDefaults_CertificateSigningRequestSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) {
 	if obj.Usages == nil {
 		obj.Usages = []certificatesv1beta1.KeyUsage{certificatesv1beta1.UsageDigitalSignature, certificatesv1beta1.UsageKeyEncipherment}
 	}
+
+	if obj.SignerName == "" {
+		// default to legacy-unknown if anything else fails
+		obj.SignerName = certificatesv1beta1.LegacyUnknownSignerName
+		csr, err := ParseCSR(&certificatesv1beta1.CertificateSigningRequest{Spec: *obj})
+		if err != nil {
+			klog.Errorf("failed to parse CSR: %v", err)
+			return
+		}
+
+		switch {
+		case isKubeletClientCSR(csr, obj.Usages):
+			obj.SignerName = certificatesv1beta1.KubeletClientSignerName
+			break
+		}
```

Comment:
> Based on my reading of the KEP, there is no defaulting behaviour for other in-built signerName - if we do want to do add them in future, we can do so here with a similar pattern

I don't anticipate any future defaulting.  This structure is ok though.

Reply:
Should we continue to sign kubelet certs with the legacy signer until they upgrade? I don't think making this defaulting logic so fancy buys us anything.

Reply:
Right now we don't have a real distinction between these two signer types, and this PR does not add support for the new signer type to kube-controller-manager (nor does it add any awareness to the existing approver & signer implementation of the `signerName` field). I don't think we've decided what the real distinction between these two `signerNames` will look like in kube-controller-manager.

I'm not sure what the implications of changing the defaulting behaviour for this field in future are though if we decide to remove defaulting the `kubelet-client` now. Would this be an acceptable change in the move from v1beta1 to v1?

Reply:
> Should we continue to sign kubelet certs with the legacy signer until they upgrade? I don't think making this defaulting logic so fancy buys us anything.

It means that a skewed kube-controller-manager is assured of doing the right thing with kubelet-client certificates on the upgrades.  It allows the legacy signer to be disconnected without upgrading kubelets immediately.  I think it's a worthwhile defaulting.

Reply:
We can't change the current behavior until we remove the v1beta1 API so this would be opt-in. On the other hand, we will have to carry this until we remove the v1beta1 API. I would not expect the capability that this enables to be widely used or useful, so I lean towards keeping it simple. We get to where we want to be in 2 releases and in the meantime, inclined operators can just upgrade their nodes. I don't feel terribly strongly about this, but I'm dubious that adding this defaulting will make a difference for anyone.

Reply:
> We can't change the current behavior until we remove the v1beta1 API so this would be opt-in.

With defaulting, given a paired kube-apiserver and kube-controller-manager, by n+2 the kube-contoller-manager never has to consider an empty signerName ever again.  Without defaulting, empty signerNames must be tolerated until we actually add a v1.

Reply:
As discussed on the call, we're going to keep this defaulting logic and require those implementing out-of-tree approval plugins to either:

1) create the appropriate RBAC role/rolebinding to allow them to continue approving CSRs
2) disable the CertificateApproval admission plugin